Feat/connect data layers

c1-ra opened the pull request:
## Description

Implement connection to SQL database, as well as methods to be used by frontend:
- `ListMetadataByUserID`
- `FindComputedStatsByID`

## Changes

`benchttp` file has been split into `report` and `computed_stats` to better separate the different structs used for different aims.


## Notes

<!-- Optional: additional notes than can help understanding the implementation -->

Commit:
Add necessary structures and interface in `benchttp` package

Commit:
Add `BenchmarkService` implementation

Commit:
Change some things to please golangci-lint

GregoryAlbouy commented on .env.example:
```diff
@@ -1,3 +1,8 @@
+HOST=host-name
+DB_USER=db-user
+DB_PASSWORD=db-password
+DB_NAME=db-name
```
[suggestion]
```suggestion
PSQL_HOST=host-name
PSQL_USER=db-user
PSQL_PASSWORD=db-password
PSQL_NAME=db-name
```

Commit:
Add `BenchmarkService` to `Server` attributes, add routes

Commit:
Add some documentation comments in `repository.go`

Commit:
Delete userId path param for `ListMetadataByUserID` for the moment

Commit:
Change `alphanum20` regex to allow only one character too

Commit:
Change `/benchmarks-list/` route to `/benchmarks/` and `benchmark/{id}` to `benchmarks/{id}`'

Commit:
Remove forgotten `objectType`, add `code5xx` to `Codestats`

Commit:
Split `report` and `benchmark` in two separate files

Commit:
Rename `benchmark` to `computedStats`

Commit:
Organize `postgresql` package into several files

c1-ra commented on cmd/main.go:
```diff
@@ -39,7 +40,12 @@ func main() {
 	}
 
-	srv := server.New(addr, rs)
+	cs, err := postgresql.NewComputedStatsService(10, 10)
```
10, 10 (max idle connections, max open connections) is totally arbitrary.
Feel free to suggest what seems appropriate, since I have no idea... 

Or we could keep default ones and not set them anywhere in the code (default are 2 and 0 (meaning unlimited)), but I feel like it is better we have control over them.
Declaring constants in main to pass them to the function could be neater though?

c1-ra replied:
While looking for some info about it, I noticed `ConnMaxIdleTime` (max time a connection can be idled before being closed) and `ConnMaxLifetime` (max time a connection can be held before being closed).
Do we want some control over them too? :thinking:

c1-ra commented on postgresql/connection.go:
```diff
@@ -0,0 +1,95 @@
+	return &computedStatsService{db}, nil
+}
+
+func (b *computedStatsService) Close() {
```
I've just realized that I have not used it anywhere :thinking:
When are my connections actually closing at the moment ? :sweat_smile:
Wouldn't it actually be better if I had a `connect()` method in `connection.go` outside `NewComputedStatsService` which would be called in each `endpoint_method` beginning? And a defer `db.Close()` in each of them? :thinking:

Commit:
Change some variable names to have more coherent ones

c1-ra commented on postgresql/connection.go:
```diff
@@ -0,0 +1,95 @@
+	return &computedStatsService{db}, nil
+}
+
+func (b *computedStatsService) Close() {
```
I have not used it anywhere. Maybe I should get rid of it?

GregoryAlbouy replied:
YES

GregoryAlbouy commented on benchttp/computed_stats.go:
```diff
@@ -0,0 +1,35 @@
+	Metadata  Metadata
+	Codestats Codestats
+	Timestats Timestats
+}
```
We'll need camelCase JSON struct tags for these, could you add them?

GregoryAlbouy commented on computed_stats.go:
```diff
@@ -0,0 +1,38 @@
+
+func (s *Server) ListMetadataByUserID(w http.ResponseWriter, r *http.Request) {
+	// TO DO: get userID from authentication to use it here instead of "1"
+	report, err := s.ComputedStatsService.ListMetadataByUserID(1)
```
Not a report!

GregoryAlbouy commented on computed_stats.go:
```diff
@@ -0,0 +1,38 @@
+		writeError(w, ErrBadRequest.Wrap(err))
+		return
+	}
+	report, err := s.ComputedStatsService.FindComputedStatsByID(idInt)
```
Not a report!

GregoryAlbouy commented on computed_stats.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func (s *Server) FindComputedStatsByID(w http.ResponseWriter, r *http.Request) {
+	idString, err := pathParam(r, idParam)
+	if err != nil {
+		writeError(w, ErrBadRequest.Wrap(err))
+		return
+	}
+
+	idInt, err := strconv.Atoi(idString)
+	if err != nil {
+		writeError(w, ErrBadRequest.Wrap(err))
+		return
+	}
```
We could extract this logic to a dedicated helper to lighten these handlers

GregoryAlbouy commented on postgresql/connection.go:
```diff
@@ -0,0 +1,95 @@
+	db *sql.DB
+}
+
+func NewComputedStatsService(idleConn, maxConn int) (benchttp.ComputedStatsService, error) {
```
It's generally preferable to return the actual struct rather than the interface it implements (although in our usage it's probably not going to be an issue)

GregoryAlbouy commented on postgresql/connection.go:
```diff
@@ -0,0 +1,95 @@
+	dbName   string
+}
+
+func getConnectionInfoFromEnvVariables() (connectionInfo, error) {
```
[suggestion] I'd recommend to handle this in `main`, and inject the retrieved values into `benchmarkService` from there: it'll make it more independent and testable!

GregoryAlbouy commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,56 @@
+func (c *computedStatsService) FindComputedStatsByID(metadataID int) (*benchttp.ComputedStats, error) {
+	computedStats := &benchttp.ComputedStats{}
+
+	stmt := "SELECT m.tag, m.finished_at, c.code_1xx, c.code_2xx, c.code_3xx, c.code_4xx, c.code_5xx, t.min, t.max, t.mean, t.median, t.variance, t.deciles " +
+		"FROM public.metadata AS m " +
+		"INNER JOIN public.codestats AS c ON c.metadata_id = m.id " +
+		"INNER JOIN public.timestats AS t ON t.metadata_id = m.id " +
+		"WHERE m.id = $1 " +
+		"ORDER BY m.finished_at DESC"
```
A bit more friendly:
```suggestion
	stmt := `
SELECT
	m.tag,
	m.finished_at,
	c.code_1xx,
	c.code_2xx,
	c.code_3xx,
	c.code_4xx,
	c.code_5xx,
	t.min,
	t.max,
	t.mean,
	t.median,
	t.variance,
	t.deciles,
FROM public.metadata AS m
INNER JOIN public.codestats AS c ON c.metadata_id = m.id
INNER JOIN public.timestats AS t ON t.metadata_id = m.id
WHERE m.id = $1
ORDER BY m.finished_at DESC`[1:]
```
Or alternatively a SQL view 🙃

GregoryAlbouy commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,56 @@
+		"ORDER BY m.finished_at DESC"
+
+	row := c.db.QueryRow(stmt, metadataID)
+	err := row.Scan(&computedStats.Metadata.Tag, &computedStats.Metadata.FinishedAt, &computedStats.Codestats.Code1xx, &computedStats.Codestats.Code2xx, &computedStats.Codestats.Code3xx, &computedStats.Codestats.Code4xx, &computedStats.Codestats.Code5xx, &computedStats.Timestats.Min, &computedStats.Timestats.Max, &computedStats.Timestats.Mean, &computedStats.Timestats.Median, &computedStats.Timestat ...
```
With appropriate struct tags (`db`), we'll probably be able to simplify this to:
```suggestion
	err := row.Scan(benchmark)
```

If for some reason we don't manage to make it work, let's at least add new lines before each field to avoid that 386 chars line!

c1-ra replied:
I was actually surprised golangci-lint did not complain about this and considered it was valid then :p
I will at least add some line returns.

GregoryAlbouy replied:
Yeah I didn't set up a linter for that because it can be an annoying constraint (if we have a long unformatted string for instance), and the last thing you want to do when you have a long line is to make it even longer by adding `//nolint` 😄

GregoryAlbouy commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/benchttp/server/benchttp"
+)
+
+func (c *computedStatsService) ListMetadataByUserID(userID int) ([]*benchttp.Metadata, error) {
```
[suggestion] Let's avoid unnecessary pointers
```suggestion
func (c *computedStatsService) ListMetadataByUserID(userID int) ([]benchttp.Metadata, error) {
```

GregoryAlbouy commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+func (c *computedStatsService) ListMetadataByUserID(userID int) ([]*benchttp.Metadata, error) {
+	metadataList := make([]*benchttp.Metadata, 0)
```
[suggestion] No need for `make` if not initialized
```suggestion
	metadataList := []benchttp.Metadata{}
```

GregoryAlbouy commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,56 @@
+	defer rows.Close()
+
+	for rows.Next() {
+		metadata := new(benchttp.Metadata)
```
[suggestion] (same about pointers)
```suggestion
		metadata := benchttp.Metadata{}
```

Commit:
Add json struct tags to `computed_stats.go`

Commit:
Add PSQL precision in `.env.example` file

Commit:
Replace `report` by `stats`

Commit:
Return struct instead of interface in `NewComputerStatsService()`
Remove unnecessary pointers
Remove `make` for slice declaration without initialization

Commit:
Add `getStringParamAndConvertToInt()` method in `computed_stats.go`

Commit:
Delete unnecessary `NewComputedStatsService.Close()` method

Commit:
Make database retrieving code more readable by adding some line returns

Commit:
Make `postgresql.computedStatsService` struct (`NewComputedStatsService()` return value) an exported type

GregoryAlbouy commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,83 @@
+func (c ComputedStatsService) FindComputedStatsByID(metadataID int) (benchttp.ComputedStats, error) {
+	computedStats := benchttp.ComputedStats{}
+
+	stmt := `SELECT
+			m.tag,
+			m.finished_at,
+			c.code_1xx,
+			c.code_2xx,
+			c.code_3xx,
+			c.code_4xx,
+			c.code_5xx,
+			t.min,
+			t.max,
+			t.mean,
+			t.median,
+			t.variance,
+			t.deciles
+		FROM public.metadata AS m
+		INNER JOIN public.codestats AS c ON c.metadata_id = m.id
+		INNER JOIN public.timestats AS t ON t.metadata_id = m.id
+		WHERE m.id = $1
+		ORDER BY m.finished_at DESC`
```
[nitpick] If this works it's not that much of a problem, but this results in a weirdly indented query, because <code>``</code> keeps the formating:
```sql
SELECT
			m.tag,
			m.finished_at,
			-- ...
		FROM public.metadata AS m
		INNER JOIN public.codestats AS c ON c.metadata_id = m.id
		-- ...
```

That's why I suggested the following:
```go
	stmt := `
SELECT
	m.tag,
	m.finished_at,
	-- ...
FROM public.metadata AS m
INNER JOIN public.codestats AS c ON c.metadata_id = m.id
-- ...`[1:]
```
(Although missing indentation inside a function can seem confusing at first, it's actually very common to do so for formatted strings)

c1-ra replied:
We tried IRL with Greg, but with the deciles and our nested structs, it is difficult :(

GregoryAlbouy commented on computed_stats.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func (s *Server) FindComputedStatsByID(w http.ResponseWriter, r *http.Request) {
+	idInt, err := getStringParamAndConvertToInt(r, idParam)
```
Actually the `ComputedStats` ID must be a string, because it matches the ID of the `Report` stored in Firestore (which is a string). Sorry for not spotting that earlier! 😕

GregoryAlbouy commented on benchttp/computed_stats.go:
```diff
@@ -0,0 +1,35 @@
+import "time"
+
+// Metadata contains a computed stats group metadata
+type Metadata struct {
+	Tag        string    `json:"tag"`
+	FinishedAt time.Time `json:"finishedAt"`
+}
```
[question] Aren't we going to need the ID on the front side?

GregoryAlbouy reviewed (commented):
Great work, only few things remaining for me!

Commit:
Reformat `FindComputedStatsByID()` query
Fix environment variable names according to `.env.example`
Add `Id` to `Metadata` struct
Change `Id` arguments from int to string

Commit:
Get environment variables in main

c1-ra commented:
@moreirathomas, we've done some modifications with @GregoryAlbouy.
We would have liked to use db struct tags for `ComputedStats`, but with our nested structs and deciles array, it seemed difficult :(
Package main will likely need some refactoring later.

Commit:
Change file, struct and method names
Regroup `firestore` and `postgresql` into `services` folder

moreirathomas commented on benchttp/repository.go:
```diff
@@ -12,3 +12,14 @@ type ReportService interface {
+
+// ComputedStatsService defines the interface to implement by a
+// service facade inside the application.
+type ComputedStatsService interface {
```
```suggestion
type StatsService interface {
```
Stats _are_ computed I think there's no need to prefix.

moreirathomas commented on benchttp/repository.go:
```diff
@@ -12,3 +12,14 @@ type ReportService interface {
+type ComputedStatsService interface {
+	// ListMetadataByUserID retrieves all the metadata (Tag and
+	// FInishedAt) related to a User with its user id.
+	ListMetadataByUserID(id string) ([]Metadata, error)
```
The description and the naming of this method is super strange.

What is `Metadata` in the context of the stats ? By what i can infer from this PR, it is some identifier for an available post-processed stats entity.

I think we can do better naming-wise.

moreirathomas commented on postgresql/connection.go:
```diff
@@ -0,0 +1,44 @@
+	_ "github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/dialers/postgres" // blank import
+)
+
+type ComputedStatsService struct {
```
```suggestion
type StatsService struct {
```

GregoryAlbouy reviewed (approved):
Amazungue PR 🚀

moreirathomas commented on postgresql/endpoint_methods.go:
```diff
@@ -0,0 +1,87 @@
+package postgresql
```
`endpoint_methods` is a strange name.

I suggest: `stats` for queries related to stats.

Implied that metadata could be renamed to better fit the fact that it deals offer things on post-processed data.

moreirathomas reviewed (commented):
My main point with the implementation is the unexpressiveness of `Metadata`.

I'm for a renaming of `Metadata` and the service/methods as a whole into something more explicit.

I suggest:

- `ComputedStatsService` -> `StatsService`
- `Metadata` -> `StatsDescriptor`
- `ComputedStatsService.ListMetadataByUserID` -> `StatsService.ListAvailable`
- `ComputedStatsService.FindComputedStatsByID` -> `StatsService.GetOne`

c1-ra commented:
I have renamed the methods, structs, fields, tables in database, and rebased on main because there were some conflicts!

moreirathomas reviewed (approved):
Nice !

Two last pet peeves, personal preference really go for what you prefer: 
- nesting `firestore` and `postgresql` inside `services` is debatable, it's more "service implementations" or "infrastructure" than pure "services", but it's nitpicking
- `benchttp.ComputedStats` could be just `benchttp.Stats`, computed is kinda obvious

GregoryAlbouy commented:
> - nesting `firestore` and `postgresql` inside `services` is debatable, it's more "service implementations" or "infrastructure" than pure "services", but it's nitpicking

Both `firestore` and `postgresql` are entirely based on a `XxxService` structure so gathering them into a `services` directory makes perfect sense to me 🤔

moreirathomas commented:
> Both `firestore` and `postgresql` are entirely based on a `XxxService` structure so gathering them into a `services` directory makes perfect sense to me 🤔

That's why I said debatable and personal pref. It's good to me, just sharing my thought on this :)

c1-ra commented:
Thank you for the reviews :) :tada:
